core/src/appleMain/kotlin/CBPeripheralCoreBluetoothPeripheral.kt

```diff
@@ -393,9 +402,47 @@ internal class CBPeripheralCoreBluetoothPeripheral(
         }
 
         val platformDescriptor = discoveredServices.obtain(descriptor)
-        return connection.execute<DidUpdateValueForDescriptor> {
+        val updatedDescriptor = connection.execute<DidUpdateValueForDescriptor> {
             centralManager.read(cbPeripheral, platformDescriptor)
-        }.descriptor.value as NSData
+        }.descriptor
+
+        // CBDescriptor "value" property type depends on cbDescriptor.UUID.UUIDString
+        // https://developer.apple.com/documentation/corebluetooth/characteristic-descriptors
+        // see type conversion table https://github.com/JuulLabs/kable/pull/706
+        when (updatedDescriptor.UUID.UUIDString) {
+            CBUUIDCharacteristicFormatString -> {
+                return updatedDescriptor.value as NSData
+            }
+            CBUUIDCharacteristicUserDescriptionString,
+            -> {
+                return (updatedDescriptor.value as String)
+                    .encodeToByteArray()
+                    .toNSData()
+            }
+            CBUUIDCharacteristicExtendedPropertiesString,
+            CBUUIDClientCharacteristicConfigurationString,
+            CBUUIDServerCharacteristicConfigurationString,
+            -> {
+                return (updatedDescriptor.value as NSNumber)
+                    .unsignedShortValue
+                    .toNSData()
+            }
+            CBUUIDL2CAPPSMCharacteristicString,
+            -> {
+                return (updatedDescriptor.value as UInt16).toNSData()
```

Comment:
The docs are saying that the `value` here is `UInt16`, but I doubt the API expects us to dereference/typecast directly as a raw pointer. It's not completely out of the question but feels a little odd for Apple to do that? My best guess is the is already an `NSData` (of two bytes length) so we should just return as `NSData` here.

I'm digging around for a device that exposes this characteristic so I can test a real example to know for sure.

Reply:
Not sure if it confirms your suspicions @davidtaylor-juul, but [this code](https://github.com/clostra/newnode/blob/6e64362fb2dd97cc58ee4b1624c466a85a43dc9f/ios/Bluetooth.m#L196-L231) retrieves the `NSData` as-is (without `UInt16` casting, as far as I can tell, at least).